mlflow/store/artifact/databricks_artifact_repo.py

```diff
@@ -320,5 +339,88 @@ def _download_file(self, remote_file_path, local_path):
+            fullpath = fullpath.rstrip("/")
+            dirpath, _ = posixpath.split(fullpath)
+            local_dir_path = os.path.join(dst_path, dirpath)
+            local_file_path = os.path.join(dst_path, fullpath)
+            if not os.path.exists(local_dir_path):
+                os.makedirs(local_dir_path)
+            return (
+                local_file_path,
+                self.thread_pool.submit(
+                    self._download_file, remote_file_path=fullpath, local_path=local_file_path
+                ),
+            )
+
+        def download_artifact_dir(dir_path):
+            """
+            Recursively list files in the directory specified by `dir_path` and submit
+            asynchronous downloads for each encountered file, returning a list of futures
+            representing each download operation.
+            """
+            download_futures = []
+            local_dir = os.path.join(dst_path, dir_path)
+            dir_content = [  # prevent infinite loop, sometimes the dir is recursively included
+                file_info
+                for file_info in self.list_artifacts(dir_path)
+                if file_info.path != "." and file_info.path != dir_path
+            ]
+            if not dir_content:  # empty dir
+                if not os.path.exists(local_dir):
+                    os.makedirs(local_dir)
+            else:
+                for file_info in dir_content:
+                    if file_info.is_dir:
+                        _, futures = download_artifact_dir(dir_path=file_info.path)
+                        download_futures += futures
+                    else:
+                        _, future = download_file(file_info.path)
+                        download_futures += [future]
+            return local_dir, download_futures
+
+        if dst_path is None:
+            dst_path = tempfile.mkdtemp()
+        dst_path = os.path.abspath(dst_path)
+
+        if not os.path.exists(dst_path):
+            raise MlflowException(
+                message=(
+                    "The destination path for downloaded artifacts does not"
+                    " exist! Destination path: {dst_path}".format(dst_path=dst_path)
+                ),
+                error_code=RESOURCE_DOES_NOT_EXIST,
+            )
+        elif not os.path.isdir(dst_path):
+            raise MlflowException(
+                message=(
+                    "The destination path for downloaded artifacts must be a directory!"
+                    " Destination path: {dst_path}".format(dst_path=dst_path)
+                ),
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+
+        if self._is_directory(artifact_path):
+            local_dir, futures = download_artifact_dir(artifact_path)
+            # Join futures to ensure that all files in the directory have
+            # been downloaded prior to returning
+            for future in futures:
+                future.result()
+            return local_dir
```

Comment:
This part can be combined with L422 if the returning signature is a dict.

In addition, see comment on L294 for util functions to handle multiple failures.

Reply:
Great idea! Addressed! (See above comments).